src/fieldlines.1
.TH "Really Slick ScreenSavers" 1 "" "X Version 11"
.de Ds
.Sp
.nf
..
.de De
.fi
..
.SH NAME
fieldlines - simulation of the electric field lines between charged particles. 
.SH SYNOPSIS
.B fieldlines
[\--root/-r]
[\--maxfps/-x \fInumber\fP]
[\--vsync/-y \fInumber\fP]
[\--dpms/-M \fInumber\fP]
[\--ions/-i \fInumber\fP]
[\--stepsize/-s \fInumber\fP]
[\--maxsteps/-m \fInumber\fP]
[\--width/-w \fInumber\fP]
[\--speed/-S \fInumber\fP]
[\--constwidth/-c]
[\--no-constwidth/-C]
[\--electric/-e]
[\--no-electric/-E]
.SH DESCRIPTION
From Terry Walsh (\fIhttp://reallyslick.com\fP):
"This is a simulation of the electric field lines between charged particles. Physics never looked so good."

Ported to Linux by Tugrul Galatali.
.SH OPTIONS
.TP 8
.B \--root
Draw on the root window.
.TP 8
.B \--maxfps \fInumber\fP
Set maximum frame rate.
.TP 8
.B \--vsync \fInumber\fP
Limit redraws to specified number of vertical refreshes.  0 - 100.  Default: 1
.TP 8
.B \--dpms \fInumber\fP
Stop rendering new frames if the display is not on.  0 - 1.  Default: 1
.TP 8
.B \--ions \fInumber\fP
Number of ions.  1 - 100.  Default: 4
.TP 8
.B \--stepsize \fInumber\fP
Step length along lines.  1 - 100.  Default: 15
.TP 8
.B \--maxsteps \fInumber\fP
Max number of steps.  1 - 1000.  Default: 100
.TP 8
.B \--width \fInumber\fP
Line width.  1 - 100.  Default: 30
.TP 8
.B \--speed \fInumber\fP
Speed.  1 - 100.  Default: 10
.TP 8
.B \--constwidth
Constant line width.  Default: False
.TP 8
.B \--electric
Electric.  Default: False
.SH ENVIRONMENT
.PP
.TP 8
.B DISPLAY
to get the default host and display number.
.SH SEE ALSO
.BR X (1),
.SH COPYRIGHT
Copyright \(co 2002 by Terry Walsh and Tugrul Galatali.  

Fieldlines is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License version 2 as
published by the Free Software Foundation.

Fieldlines is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.SH AUTHOR
Terry Walsh and Tugrul Galatali <tugrul@galatali.com>



ChangeLog
******************************* Version 0.9.1  *********************************

	Added port of Pixel City, originally by Shamus Young,
http://code.google.com/p/pixelcity/
	Replaced methods deprecated by ImageMagick.

******************************* Version 0.9.0  *********************************

	Added port of drempels, originally by Ryan M. Geiss,
http://www.geisswerks.com/about_drempels.html
	Added new saver feedback by Tugrul Galatali.
	Added support for DPMS to idle the saver when the display isn't on.
	Fixed brown paper bag bug in hyperspace.
	Added two new presets to flocks inspired by Matt Richard's Wavy Gravy on
openprocessing.org
	Tweaked hufo_smoke to avoid flat flame effect.
	Fixed unknown argument segfault in lattice.
	Switched matrixview's main character array to vertex arrays, which helps
significantly on older nVidia hardware.
	hyperspace and drempels have command line arguments to disable use of
shaders in case they behave unexpectedly.
	Minor changes to fix builds on other platforms.

******************************* Version 0.8.3 *********************************

	New screen saver Lorenz contributed by Soeren Sonnenburg <sonne@debian.org>
	Fix for SF #2673397 by changing =='s to >='s in widthChange update logic,
otherwise non-sense values for interpolation would wreak havoc. Also changed for
xyzChange.

******************************* Version 0.8.2 *********************************

	Limit frame rate to vertical refresh by default if compiled with GLEW
and GLX_SGI_swap_control is available. Otherwise, maxfps set to 30.
	--nice eliminated and made default. nanosleep/usleep are much better 
behaved these days.
	spirographx modified to use vertex arrays and GNU extension sincosf if
available, dropping CPU usage significantly.
	Minor updates to correct issues noted by newer compilers.
	--disable-debug is now the default in configure.

******************************* Version 0.8.1 *********************************

	Pulled in vroot.h patch from Ari Pollak's Debian package.
	Added missing #include <assert.h> to hyperspace.cpp
	Image loading in matrixview was hanging for whatever reason,
		rewritten with the new Magick Wand interface.
	ImageMagick now required.
	timeval to double conversion now fixed with an explicit cast of tv_sec
		to double.
	Fixed automake files to use CLEANFILES
	Dropped SSE/3dnow stuff
	Fixed hyperspace.xml typo in dist_kdessconfig_DATA

******************************* Version 0.8.0 *********************************

	Rough sync against new Windows release, rssavers 0.1
		Except for skyrocket
		Added new saver Hyperspace

******************************* Version 0.7.6 *********************************

	Fix to matrixview.c where the body of a thread was accidentally left as a
subfunction of a function that exits while the thread is still alive.

	Added an autoconf check for ExportImagePixels since it was very recently
added (5.5.7).

******************************* Version 0.7.5 *********************************

	Removed multiline strings that are no longer supported by gcc 3.3.

	Added spirographx, a port of an OSX app/screensaver SpirographX by 
Holmes Futrell <holmes@neatosoftware.com>.
	http://homepage.mac.com/solidmag/spiro/

	Added matrixview, based on Alex Zolotov's Matrix GL. Hacked in external 
image loading and ability to modify the resolution of the matrix.

******************************* Version 0.7.4 *********************************

	Fixed bug in lattice when the variable passed to acos would be slightly
greater than 1, causing cascading nan's that made the program useless.

	Explicitly clearing buffers to avoid garbage in blurring modes.

	Lots of stupid little bug fixes.

******************************* Version 0.7.3 *********************************

	Added trails and blur options to flocks. Also added presets to spruce
things up in combination with random preset selection when running 
unconfigured.

	Fixed another cpu detection bug, damn off by one errors :)

	Added support for decimal integer arguments to -window-id. 

	Added .desktop files for preliminary support of KDE. Not sure where
the default directory is, but if it isn't specified by --with-kdessconfigdir,
the configure script checks for the existence of:

/usr/share/applnk/System/ScreenSavers

******************************* Version 0.7.2 *********************************

	Fixed bug in cpu detection code.

	Fixed a bug that included audio data in skyrocket when OpenAL wasn't 
found.

******************************* Version 0.7.1 *********************************

	Fixed a typo crashing skyrocket on big endian machines. Spotted by
Michel Danzer.

	Two new savers: biof and busyspheres.

	New texture smooth for colorfire.

	Fixed a bug that left screensavers running after they should have quit.

	Replaced compile time cpu feature detection with runtime detection.

******************************* Version 0.7.0 *********************************

	Extracted all binary data from the source into binary files. The
source distribution has gotten notably smaller as a result. 

	The binary files are now converted during compilation to significantly 
larger source files with escaping instead of base64 type inefficient strings.
The resulting executables have no unnecessary bloat.

	If the bzip2 library is detected on the system, it is used to compress
the binary data before conversion.

	Added --disable-bzip2 flag to configure. This is to give up space in
order to avoid a delay while decompressing, which is noticeable in skyrocket
that has about 3MB of compressed textures and audio.

	Added --disable-sound flag to configure. This chops off about 3.3MB
(before compression) of audio from skyrocket and any runtime overhead
associated with sound. Useful if you intend to keep passing volume = 0 to
skyrocket.

	Added SSE/3DNow optimized code that gives euphoria a nice 10% boost
on my P4, probably a bit more on an Athlon but I haven't had recent console 

src/fieldlines.cpp
/*
 * Copyright (C) 2002  Terence M. Welsh
 * Ported to Linux by Tugrul Galatali <tugrul@galatali.com>
 *
 * Field Lines is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as 
 * published by the Free Software Foundation.
 *
 * Field Lines is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

// Field Lines screensaver

#include <math.h>
#include <stdio.h>
#include <GL/gl.h>
#include <GL/glu.h>

#include "driver.h"
#include "rsDefines.h"
#include "rsRand.h"

const char *hack_name = "fieldlines";

float wide, high, deep;

class ion;

// Globals
int readyToDraw = 0;
ion *ions;
float aspectRatio;
float elapsedTime;

// Parameters edited in the dialog box
int dIons;
int dStepSize;
int dMaxSteps;
int dWidth;
int dSpeed;
int dConstwidth;
int dElectric;

class ion {
      public:
	float charge;
	float xyz[3];
	float vel[3];
	float angle;
	float anglevel;

	  ion ();
	 ~ion () {};
	void update ();
};

ion::ion ()
{
	charge = float (rsRandi (2));

	if (rsRandi(2))
		charge = -1.0f;
	else
		charge = 1.0f;

	xyz[0] = rsRandf (2.0f * wide) - wide;
	xyz[1] = rsRandf (2.0f * high) - high;
	xyz[2] = rsRandf (2.0f * deep) - deep;
	vel[0] = rsRandf (float (dSpeed) * 4.0f) - (float (dSpeed) * 2.0f);
	vel[1] = rsRandf (float (dSpeed) * 4.0f) - (float (dSpeed) * 2.0f);
	vel[2] = rsRandf (float (dSpeed) * 4.0f) - (float (dSpeed) * 2.0f);

	angle = 0.0f;
	anglevel = 0.0005f * float (dSpeed) + 0.0005f * rsRandf (float (dSpeed));
}

void
  ion::update ()
{
	xyz[0] += vel[0] * elapsedTime;
	xyz[1] += vel[1] * elapsedTime;
	xyz[2] += vel[2] * elapsedTime;
	if (xyz[0] > wide)
		vel[0] -= 0.1f * float (dSpeed);

	if (xyz[0] < -wide)
		vel[0] += 0.1f * float (dSpeed);

	if (xyz[1] > high)
		vel[1] -= 0.1f * float (dSpeed);

	if (xyz[1] < -high)
		vel[1] += 0.1f * float (dSpeed);

	if (xyz[2] > deep)
		vel[2] -= 0.1f * float (dSpeed);

	if (xyz[2] < -deep)
		vel[2] += 0.1f * float (dSpeed);

	angle += anglevel;
	if (angle > PIx2)
		angle -= PIx2;
}

void drawfieldline (int source, float x, float y, float z)
{
	int i, j;
	float charge;
	float repulsion;
	float dist, distsquared, distrec;
	float xyz[3];
	float lastxyz[3];
	float dir[3];
	float end[3];
	float tempvec[3];
	float r, g, b;
	float lastr, lastg, lastb;
	static float brightness = 10000.0f;

	charge = ions[source].charge;
	lastxyz[0] = ions[source].xyz[0];
	lastxyz[1] = ions[source].xyz[1];
	lastxyz[2] = ions[source].xyz[2];
	dir[0] = x;
	dir[1] = y;
	dir[2] = z;

	// Do the first segment
	r = float (fabs (dir[2])) * brightness;
	g = float (fabs (dir[0])) * brightness;
	b = float (fabs (dir[1])) * brightness;

	if (r > 1.0f)
		r = 1.0f;
	if (g > 1.0f)
		g = 1.0f;
	if (b > 1.0f)
		b = 1.0f;
	lastr = r;
	lastg = g;
	lastb = b;
	glColor3f (r, g, b);
	xyz[0] = lastxyz[0] + dir[0];
	xyz[1] = lastxyz[1] + dir[1];
	xyz[2] = lastxyz[2] + dir[2];
	if (dElectric) {
		xyz[0] += rsRandf (float (dStepSize) * 0.2f) - (float (dStepSize) * 0.1f);
		xyz[1] += rsRandf (float (dStepSize) * 0.2f) - (float (dStepSize) * 0.1f);
		xyz[2] += rsRandf (float (dStepSize) * 0.2f) - (float (dStepSize) * 0.1f);
	}
	if (!dConstwidth)
		glLineWidth ((xyz[2] + 300.0f) * 0.000333f * float (dWidth));

	glBegin (GL_LINE_STRIP);
	glColor3f (lastr, lastg, lastb);
	glVertex3fv (lastxyz);
	glColor3f (r, g, b);
	glVertex3fv (xyz);
	if (!dConstwidth)
		glEnd ();

	for (i = 0; i < int (dMaxSteps); i++) {
		dir[0] = 0.0f;
		dir[1] = 0.0f;
		dir[2] = 0.0f;
		for (j = 0; j < int (dIons); j++) {
			repulsion = charge * ions[j].charge;
			tempvec[0] = xyz[0] - ions[j].xyz[0];
			tempvec[1] = xyz[1] - ions[j].xyz[1];
			tempvec[2] = xyz[2] - ions[j].xyz[2];
			distsquared = tempvec[0] * tempvec[0] + tempvec[1] * tempvec[1] + tempvec[2] * tempvec[2];
			dist = float (sqrt (distsquared));

			if (dist < float (dStepSize) && i > 2) {
				end[0] = ions[j].xyz[0];
				end[1] = ions[j].xyz[1];
				end[2] = ions[j].xyz[2];
				i = 10000;
			}
			tempvec[0] /= dist;
			tempvec[1] /= dist;
			tempvec[2] /= dist;
			if (distsquared < 1.0f)
				distsquared = 1.0f;
			dir[0] += tempvec[0] * repulsion / distsquared;
			dir[1] += tempvec[1] * repulsion / distsquared;
			dir[2] += tempvec[2] * repulsion / distsquared;
		}
		lastr = r;
		lastg = g;
		lastb = b;
		r = float (fabs (dir[2])) * brightness;
		g = float (fabs (dir[0])) * brightness;
		b = float (fabs (dir[1])) * brightness;

		if (dElectric) {
			r *= 10.0f;
			g *= 10.0f;
			b *= 10.0f;;
			if (r > b * 0.5f)
				r = b * 0.5f;
			if (g > b * 0.3f)
				g = b * 0.3f;
		}
		if (r > 1.0f)
			r = 1.0f;
		if (g > 1.0f)
			g = 1.0f;
		if (b > 1.0f)
			b = 1.0f;
		distsquared = dir[0] * dir[0] + dir[1] * dir[1] + dir[2] * dir[2];
		distrec = float (dStepSize) / float (sqrt (distsquared));

		dir[0] *= distrec;
		dir[1] *= distrec;
		dir[2] *= distrec;
		if (dElectric) {
			dir[0] += rsRandf (float (dStepSize)) - (float (dStepSize) * 0.5f);
			dir[1] += rsRandf (float (dStepSize)) - (float (dStepSize) * 0.5f);
			dir[2] += rsRandf (float (dStepSize)) - (float (dStepSize) * 0.5f);
		}
		lastxyz[0] = xyz[0];
		lastxyz[1] = xyz[1];
		lastxyz[2] = xyz[2];
		xyz[0] += dir[0];
		xyz[1] += dir[1];
		xyz[2] += dir[2];
		if (!dConstwidth) {
			glLineWidth ((xyz[2] + 300.0f) * 0.000333f * float (dWidth));

			glBegin (GL_LINE_STRIP);
		}
		glColor3f (lastr, lastg, lastb);
		glVertex3fv (lastxyz);
		if (i != 10000) {
			if (i == (int (dMaxSteps) - 1))
				glColor3f (0.0f, 0.0f, 0.0f);
			else
				glColor3f (r, g, b);
			glVertex3fv (xyz);
			if (i == (int (dMaxSteps) - 1))
				glEnd ();
		}
	}
	if (i == 10001) {
		glColor3f (r, g, b);
		glVertex3fv (end);
		glEnd ();
	}
}

void hack_draw (xstuff_t * XStuff, double currentTime, float frameTime)
{
	int i;
	static float s = float (sqrt (float (dStepSize) * float (dStepSize) * 0.333f));

	elapsedTime = frameTime;

	glClear (GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

	for (i = 0; i < dIons; i++)
		ions[i].update ();

	for (i = 0; i < dIons; i++) {
		drawfieldline (i, s, s, s);
		drawfieldline (i, s, s, -s);
		drawfieldline (i, s, -s, s);
		drawfieldline (i, s, -s, -s);
		drawfieldline (i, -s, s, s);
		drawfieldline (i, -s, s, -s);
		drawfieldline (i, -s, -s, s);
		drawfieldline (i, -s, -s, -s);
	}
}

void hack_reshape (xstuff_t * XStuff)
{
	// Window initialization
	glViewport (0, 0, XStuff->windowWidth, XStuff->windowHeight);

	aspectRatio = float (XStuff->windowWidth) / float (XStuff->windowHeight);
	if (XStuff->windowWidth > XStuff->windowHeight) {
		high = deep = 160.0f;
		wide = high * aspectRatio;
	} else {
		wide = deep = 160.0f;
		high = wide * aspectRatio;
	}

	glEnable (GL_DEPTH_TEST);
	glEnable (GL_LINE_SMOOTH);
	glClearColor (0.0f, 0.0f, 0.0f, 1.0f);

	glMatrixMode (GL_PROJECTION);
	glLoadIdentity ();
	gluPerspective (60.0, aspectRatio, 50, 3000);

	glMatrixMode (GL_MODELVIEW);
	glLoadIdentity ();
	glTranslatef (0.0, 0.0, -2 * deep);
}

void hack_init (xstuff_t * XStuff)
{
	hack_reshape (XStuff);

	if (dConstwidth)
		glLineWidth (float (dWidth) * 0.1f);

	ions = new ion[dIons];
}

void hack_cleanup (xstuff_t * XStuff)
{
}

void hack_handle_opts (int argc, char **argv)
{
	dIons = 6;
	dStepSize = 10;
	dMaxSteps = 300;
	dWidth = 30;
	dSpeed = 10;
	dConstwidth = 0;
	dElectric = 0;

	while (1) {
		int c;

#ifdef HAVE_GETOPT_H
		static struct option long_options[] = {
			{"help", 0, 0, 'h'},
			DRIVER_OPTIONS_LONG {"ions", 1, 0, 'i'},
			{"stepsize", 1, 0, 's'},
			{"maxsteps", 1, 0, 'm'},
			{"width", 1, 0, 'w'},
			{"speed", 1, 0, 'S'},
			{"constwidth", 0, 0, 'c'},
			{"no-constwidth", 0, 0, 'C'},
			{"electric", 0, 0, 'e'},
			{"no-electric", 0, 0, 'E'},
			{0, 0, 0, 0}
		};

		c = getopt_long (argc, argv, DRIVER_OPTIONS_SHORT "hi:s:m:w:S:cCeE", long_options, NULL);
#else
		c = getopt (argc, argv, DRIVER_OPTIONS_SHORT "hrx:i:s:m:w:S:cCeE");
#endif
		if (c == -1)
			break;

		switch (c) {
			DRIVER_OPTIONS_CASES case 'h':printf ("%s:"
#ifndef HAVE_GETOPT_H
							      " Not built with GNU getopt.h, long options *NOT* enabled."
#endif
							      "\n" DRIVER_OPTIONS_HELP "\t--ions/-i <arg>\n" "\t--stepsize/-s <arg>\n" "\t--maxsteps/-m <arg>\n"
							      "\t--width/-w <arg>\n" "\t--speed/-S <arg>\n" "\t--constwidth/-c\n" "\t--no-constwidth/-C\n" "\t--electric/-e\n"
							      "\t--no-electric/-E\n", argv[0]);
			exit (1);
		case 'i':
			dIons = strtol_minmaxdef (optarg, 10, 1, 100, 1, 4, "--ions: ");
			break;
		case 's':
			dStepSize = strtol_minmaxdef (optarg, 10, 1, 100, 1, 15, "--stepsize: ");
			break;
		case 'm':
			dMaxSteps = strtol_minmaxdef (optarg, 10, 1, 1000, 1, 100, "--maxsteps: ");
			break;
		case 'w':
			dWidth = strtol_minmaxdef (optarg, 10, 1, 100, 1, 30, "--width: ");
			break;
		case 'S':
			dSpeed = strtol_minmaxdef (optarg, 10, 1, 100, 1, 10, "--speed: ");
			break;
		case 'c':
			dConstwidth = 1;
			break;
		case 'C':
			dConstwidth = 0;
			break;
		case 'e':
			dElectric = 1;
			break;
		case 'E':
			dElectric = 0;
			break;
		}
	}
}
